File: ui/src/routes/api/gaps.ts
import { defineApi } from '@/server/api-route'
import { json } from '@/server/http'
import { requireUser } from '@/server/api-guard'
import { listGaps } from '@/server/gaps'

// The Studio's Suggested queue: capability gaps agents have reported, ranked
// by how often the work-shape recurs. Any member reads (the queue is what
// invites people to tailor); status changes live on /api/gaps/$id.
export const Route = defineApi('/api/gaps', {
  GET: async ({ request }) => {
    const user = await requireUser(request)
    if (user instanceof Response) return user
    const url = new URL(request.url)
    const status = url.searchParams.get('status') ?? undefined
    return json({ gaps: await listGaps(status) })
  },
})

File: api/src/routes/agents/gaps.rs
// /api/gaps. GET → the Studio's Suggested queue: capability gaps agents have reported,
// ranked by how often the work-shape recurs. Any member reads (the queue is
// what invites people to tailor); status changes live on /api/gaps/{id}.

use crate::gaps::list_gaps;
use crate::session::require_user;
use crate::state::AppState;
use axum::Json;
use axum::extract::{Query, State};
use axum::http::HeaderMap;
use axum::response::{IntoResponse, Response};
use serde_json::json;

#[derive(serde::Deserialize)]
pub struct GapsQuery {
    status: Option<String>,
}

pub async fn get(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(query): Query<GapsQuery>,
) -> Response {
    if let Err(gate) = require_user(&state, &headers).await {
        return gate;
    }
    match list_gaps(&state.pg, query.status.as_deref()).await {
        Ok(gaps) => Json(json!({ "gaps": gaps })).into_response(),
        Err(e) => {
            tracing::error!("[gaps] list failed: {e}");
            crate::error::thrown_internal_error()
        }
    }
}
